apps/cms.c

```diff
@@ -877,23 +877,31 @@ int cms_main(int argc, char **argv)
         for (i = 0; i < sk_X509_num(encerts); i++) {
             CMS_RecipientInfo *ri;
             cms_key_param *kparam;
-            int tflags = flags;
+            int tflags = flags | CMS_KEY_PARAM; /* This flag enforces allocating the EVP_PKEY_CTX for the recipient here */
```

Comment:
This top bit is not too different as it was doing this most of the time anyway.
Looks like the KeyAgree path should also work with this change as it doesnt use the flag.

Reply:
Here is the place where we process this bit:

https://github.com/openssl/openssl/blob/f64f26220442db3c6913188e6014e5bc5bc34653/crypto/cms/cms_env.c#L164-L171

If it is not set, we just call the default envelope ASN1 control. If it is, we first initialize the ctx, then set params and call the envelope ASN1 control that can refer to set parameters.